client/src/components/search/with-instant-search.tsx

```diff
@@ -0,0 +1,207 @@
+      };
+
+const mapStateToProps = createSelector(
+  searchQuerySelector,
+  isSearchDropdownEnabledSelector,
+  function (
+    query: string,
+    isDropdownEnabled: boolean
+  ): { query: string; isDropdownEnabled: boolean } {
+    return { query, isDropdownEnabled };
+  }
+);
+const mapDispatchToProps = {
+  toggleSearchDropdown,
+  updateSearchQuery
+};
+
+function searchStateToUrl(
+  { pathname }: { pathname: string },
+  query: string
+): string {
+  return `${pathname}${query ? `?${qs.stringify({ query })}` : ''}`;
+}
+function urlToSearchState({ search }: WindowLocation): { query: string } {
+  return qs.parse(search.slice(1)) as { query: string };
+}
+
+interface InstantSearchRootProps {
+  isDropdownEnabled: boolean;
+  location: WindowLocation<{ query: string }>;
+  query: string;
+  toggleSearchDropdown: (toggle: boolean) => void;
+  updateSearchQuery: (query: string) => void;
+  children: ReactNode;
+}
+function InstantSearchRoot({
+  isDropdownEnabled,
+  location,
+  query,
+  toggleSearchDropdown,
+  updateSearchQuery,
+  children
+}: InstantSearchRootProps) {
+  function isSearchPage(): boolean {
+    return location.pathname.startsWith('/search');
+  }
+
+  function setQueryFromURL(): void {
+    if (isSearchPage()) {
+      const { query: queryFromURL } = urlToSearchState(location);
+      if (query !== queryFromURL) {
+        updateSearchQuery(queryFromURL);
+      }
+    }
+  }
+
+  useEffect(() => {
+    toggleSearchDropdown(!isSearchPage());
+    setQueryFromURL();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  // https://reactjs.org/docs/hooks-faq.html#how-to-get-the-previous-props-or-state
+  const prevLocationRef = useRef<InstantSearchRootProps['location']>();
+  useEffect(() => {
+    prevLocationRef.current = location;
+  });
+  const prevLocation = prevLocationRef.current;
+  useEffect(() => {
+    const enableDropdown = !isSearchPage();
+    if (isDropdownEnabled !== enableDropdown) {
+      toggleSearchDropdown(enableDropdown);
+    }
+
+    if (location !== prevLocation) {
+      if (isSearchPage()) {
+        if (
+          location.state &&
+          typeof location.state === 'object' &&
+          location.state.hasOwnProperty('query')
+        ) {
+          updateSearchQuery(location.state.query);
+        } else if (location.search) {
+          setQueryFromURL();
+        } else {
+          void navigate(searchStateToUrl(location, query), {
+            state: { query },
+            replace: true
+          });
+        }
+      } else if (query) {
+        updateSearchQuery('');
+      }
+    }
+  });
+
+  const debouncedSetState = useRef<ReturnType<typeof setTimeout>>(
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    setTimeout(() => {})
+  );
+  function updateBrowserHistory(query: string): void {
+    if (isSearchPage()) {
+      clearTimeout(debouncedSetState.current);
+
+      debouncedSetState.current = setTimeout(() => {
```

Comment:
```suggestion
      debouncedSetState.current = window.setTimeout(() => {
```
If we use `number` above, we have to be explicit.